libsolidity/codegen/Compiler.cpp

```diff
@@ -37,14 +37,16 @@ void Compiler::compileContract(
 	bytes const& _metadata
 )
 {
-	auto static notTransient = [](VariableDeclaration const* _varDeclaration) {
+	auto static isTransientReferenceType = [](VariableDeclaration const* _varDeclaration) {
 		solAssert(_varDeclaration);
-		return _varDeclaration->referenceLocation() != VariableDeclaration::Location::Transient;
+		return
+			_varDeclaration->referenceLocation() == VariableDeclaration::Location::Transient &&
+			!_varDeclaration->type()->isValueType();
 	};
 
 	solUnimplementedAssert(
-		ranges::all_of(_contract.stateVariables(), notTransient),
-		"Transient storage variables are not supported."
+		ranges::none_of(_contract.stateVariables(), isTransientReferenceType),
+		"Transient storage reference type variables are not supported."
```

Comment:
There's one more thing we should be rejecting - pre-Cancun EVMs. The same way we report an error if you try to use e.g. `block.basefee` or `shl()` opcode while selecting an EVM that does not have it.

That would be a regular error, rather than unimplemented error though (it cannot be implemented).